applications/GeoMechanicsApplication/tests/cpp_tests/test_time_loop_executor.cpp

```diff
@@ -0,0 +1,356 @@
+{
+    TimeLoopExecutor executor;
+    const auto increments = std::vector<double>{0.1, 0.2, 0.3, 0.4};
+    executor.SetTimeIncrementor(std::make_unique<PrescribedTimeIncrementor>(increments));
+    executor.SetSolverStrategyTimeStepExecutor(
+            std::make_unique<DummySolverStrategy>(TimeStepEndState::ConvergenceState::converged));
+    const auto step_states = executor.Run(MakeConvergedStepState());
+
+    KRATOS_EXPECT_EQ(increments.size(), step_states.size());
+    KRATOS_EXPECT_TRUE(std::all_of(step_states.begin(), step_states.end(),
+                       [](const auto& state) {return state.num_of_cycles == 1;}))
+    auto time = 0.0;
+    for (auto index = std::size_t{0}; index < increments.size(); ++index) {
+        time += increments[index];
+        KRATOS_EXPECT_DOUBLE_EQ(time, step_states[index].time);
+    }
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ExpectEndTimeToBeSetOnSolverStrategyAfterRunningAStepLoop, KratosGeoMechanicsFastSuite)
+{
+    auto solver_strategy = RunFixedCycleTimeLoop(1);
+
+    KRATOS_EXPECT_DOUBLE_EQ(1.0, solver_strategy->GetEndTime());
+    KRATOS_EXPECT_DOUBLE_EQ(0.5, solver_strategy->GetTimeIncrement());
+    KRATOS_EXPECT_EQ(2, solver_strategy->GetStepNumber());
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ExpectATimeStepCloneAtStartOfStep, KratosGeoMechanicsFastSuite)
+{
+    auto solver_strategy = RunFixedCycleTimeLoop(1);
+
+    KRATOS_EXPECT_TRUE(solver_strategy->IsCloned())
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ExpectNoRestoreCalledForOneCycle, KratosGeoMechanicsFastSuite)
+{
+    auto solver_strategy = RunFixedCycleTimeLoop(1);
+
+    KRATOS_EXPECT_FALSE(solver_strategy->IsRestoreCalled())
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ExpectRestoreCalledForTwoCycles, KratosGeoMechanicsFastSuite)
+{
+    auto solver_strategy = RunFixedCycleTimeLoop(2);
+
+    KRATOS_EXPECT_TRUE(solver_strategy->IsRestoreCalled())
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ExpectDisplacementFieldStoredForResetDisplacements, KratosGeoMechanicsFastSuite)
+{
+    auto solver_strategy = RunFixedCycleTimeLoop(1);
+
+    KRATOS_EXPECT_TRUE(solver_strategy->IsSaveFieldCalled())
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ExpectDisplacementFieldUpdateForEveryStep, KratosGeoMechanicsFastSuite)
+{
+    auto solver_strategy = RunFixedCycleTimeLoop(1);
+
+    KRATOS_EXPECT_EQ(2, solver_strategy->GetCountAccumulateTotalDisplacementFieldCalled());
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ExpectOutputProcessCalledForEveryStep, KratosGeoMechanicsFastSuite)
+{
+    TimeLoopExecutor executor;
+    const auto wanted_num_of_cycles_per_step = 1;
+    executor.SetTimeIncrementor(std::make_unique<FixedCyclesTimeIncrementor>(wanted_num_of_cycles_per_step));
+    auto solver_strategy = std::make_shared<DummySolverStrategy>(TimeStepEndState::ConvergenceState::converged);
+    executor.SetSolverStrategyTimeStepExecutor(solver_strategy);
+    const auto step_states = executor.Run(TimeStepEndState{});
+    KRATOS_EXPECT_EQ(step_states.size(), solver_strategy->GetCountOutputProcessCalled());
```

Comment:
I would add a check for the number of expected step states. Suppose that for whatever reason it no longer returns two states, but only one, and that consequently the output generation process is only called once, this test would not fail. Something along the lines of:
```
KRATOS_EXPECT_EQ(2, step_states.size());
```

Reply:
Done